crates/exex/exex/src/dyexex.rs

```diff
@@ -0,0 +1,39 @@
+//! Type-safe abstractions for Dynamically Loaded ExExes
+
+/// Dynamically loads an ExEx entrypoint, which accepts a user-defined function representing the
+/// core ExEx logic. The provided function must take an [`ExExContext`](`crate::ExExContext`) as its
+/// argument.
+///
+/// # Returns
+/// A Future that will be polled by the [`ExExManager`](`crate::ExExManager`).
+///
+/// ## Example usage:
+/// ```rust
+/// use reth_exex::{define_exex, ExExContext};
+/// use reth_node_api::FullNodeComponents;
+/// use std::future::Future;
+///
+/// // Create a function to produce ExEx logic
+/// async fn exex<Node: FullNodeComponents>(
+///     _ctx: ExExContext<Node>,
+/// ) -> eyre::Result<impl std::future::Future<Output = eyre::Result<()>>> {
+///     let _exex = async move { Ok(()) };
+///     Ok(_exex)
+/// }
+///
+/// // Use the macro to generate the entrypoint function
+/// define_exex!(exex);
+/// ```
+#[macro_export]
+macro_rules! define_exex {
+    ($user_fn:ident) => {
+        #[no_mangle]
+        pub extern "C" fn _launch_exex<Node: FullNodeComponents>(
```

Comment:
it shouldn't be a C-ABI extern, we need to either omit the string or specify "Rust" explicitly

C ABI will not allow us to return a Future

Reply:
you mean not allow us to pass closure into `_launch_exex ` / macro?

Reply:
we still can load like that

```rust
    unsafe fn load_one<R, Node>(
        &self,
        path: impl AsRef<Path>,
        ctx: ExExContext<Node>,
    ) -> Result<Box<dyn FnOnce(ExExContext<Node>) -> R + Send>>
    where
        R: Future<Output = Result<()>>,
        Node: FullNodeComponents,
    {
        type Func<R, Node> =
            unsafe fn(ExExContext<Node>) -> *mut (dyn FnOnce(ExExContext<Node>) -> R + Send);

        let lib = libloading::Library::new(path.as_ref())?;
        let raw_func_pointer: libloading::Symbol<'_, Func<R, Node>> = lib.get(USER_FN_NAME)?;

        let exex: Box<dyn FnOnce(ExExContext<Node>) -> R + Send> =
            Box::from_raw(raw_func_pointer(ctx));
        mem::forget(lib);

        Ok(exex)
    }
```

Reply:
@shekhirin can you clarify what kind of futures usage we missed with “C”?